game/log.cc

```diff
@@ -37,159 +38,172 @@
-		close(fd[1]);  // Close the original pipe begin
+	/// Duplicate the real stderr into another fd and bind that to boost iostream, restore on exit
+	struct Stream: boost::iostreams::stream<boost::iostreams::file_descriptor_sink> {
+		Stream(): stream(dup(STDERR_FILENO), boost::iostreams::close_handle) {}
+		~Stream() { (void)dup2((*this)->handle(), STDERR_FILENO); }
+	} stream;
+	StreamRedirect redirect{ std::cerr, stream.rdbuf() };  // Make std::cerr output to our copy of real stderr
+	StderrGrabber() {
+		Pipe pipefd;
+		// Replace STDERR_FILENO with a duplicate of pipefd.send
+		if (dup2(pipefd.send, STDERR_FILENO) == -1) throw std::runtime_error("dup2()");
 		std::clog << "stderr/info: Standard error output redirected here\n" << std::flush;
-		logger = std::async(std::launch::async, [fdpipe = fd[0]] {
+		// Start a thread that logs messages from pipe; this is terminated when ~Stream restores STDERR_FILENO (closing the sending end)
+		logger = std::async(std::launch::async, [fd = std::move(pipefd.recv)]() mutable {
 			std::string line;
 			unsigned count = 0;
-			for (char ch; read(fdpipe, &ch, 1) == 1;) {
+			for (char ch; read(fd, &ch, 1) == 1;) {
 				line += ch;
 				if (ch != '\n') continue;
 				std::clog << "stderr/info: " + line << std::flush;
 				line.clear(); ++count;
 			}
-			close(fdpipe);  // Close this end of pipe
-			if (count > 0) std::clog << "stderr/notice: " << count << " messages logged to stderr/info\n" << std::flush;
+			if (count > 0) std::clog << "stderr/notice: " + std::to_string(count) + " messages logged to stderr/info\n" << std::flush;
 		});
 	}
-	~StderrGrabber() {
-		dup2(stream->handle(), STDERR_FILENO);  // Restore stderr (closes the pipe, terminating the thread)
-		std::cerr.rdbuf(backup);  // Restore original rdbuf (that writes to normal stderr)
-	}
+	~StderrGrabber() { close(STDERR_FILENO); }
 };
 #else
 struct StderrGrabber {};  // Not supported on Windows
 #endif
 
-std::unique_ptr<StderrGrabber> grabber;
-
-/** \internal The implementation of the stream filter that handles the message filtering. **/
-class VerboseMessageSink : public boost::iostreams::sink {
-  public:
+/// Receives strings written to boost::iostreams::stream_buffer<VerboseMessageSink>
+struct VerboseMessageSink: boost::iostreams::sink {
 	std::streamsize write(const char* s, std::streamsize n);
 };
 
-// defining them in main() causes segfault at exit as they apparently got free'd before we're done using them
-static boost::iostreams::stream_buffer<VerboseMessageSink> sb; //!< \internal
-static VerboseMessageSink vsm; //!< \internal
-
-//! \internal used to store the default/original clog buffer.
-static std::streambuf* default_ClogBuf = nullptr;
-
-fs::ofstream file;
-
-std::string target;
-int minLevel;
-
-void writeLog(std::string const& msg) {
-	std::lock_guard<std::mutex> l(log_lock);
-	std::cerr << msg << std::flush;
-	file << msg << std::flush;
-}
+static struct Impl {
+	const std::map<std::string, int> numeric = {
```

Comment:
Personally i'd rename this to `loglevels` or something like that. Numeric is pretty much not saying anything